analyzers/src/SonarAnalyzer.Common/Rules/UseWhereBeforeOrderByBase.cs

```diff
@@ -0,0 +1,71 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class UseWhereBeforeOrderByBase<TSyntaxKind, TInvocation> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+    where TInvocation : SyntaxNode
+{
+    private const string DiagnosticId = "S6607";
+    protected override string MessageFormat => "\"Where\" should be used before \"{0}\"";
+
+    protected abstract bool TryGetOperands(TInvocation invocation, out SyntaxNode left, out SyntaxNode right);
+
+    protected UseWhereBeforeOrderByBase() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(Language.GeneratedCodeRecognizer, c =>
+        {
+            var invocation = (TInvocation)c.Node;
+
+            if (Language.GetName(invocation).Equals("Where", Language.NameComparison)
+                && TryGetOperands(invocation, out var left, out var right)
+                && LeftHasCorrectName(left, out var orderByMethodDescription)
+                && MethodIsLinqExtension(left, c.SemanticModel)
+                && MethodIsLinqExtension(right, c.SemanticModel))
+            {
+                var diagnostic = Diagnostic.Create(
+                    Rule,
+                    Language.GetIdentifier(right)?.GetLocation(),
+                    new[] { Language.GetIdentifier(left)?.GetLocation() },
+                    orderByMethodDescription);
+                c.ReportIssue(diagnostic);
+            }
+        },
+        Language.SyntaxKind.InvocationExpression);
+
+    private bool LeftHasCorrectName(SyntaxNode left, out string methodName)
```

Comment:
And what about cases like these?
```
list.OrderBy(x => x).Append(1).Where(x => false);
list.OrderBy(x => x).AsEnumerable().Where(x => false);
```
Don't get me wrong, I think we can live with FNs, especially if fixing them will complicate a lot the implementation (and they are not that frequent either). I was just wondering if those are indeed FNs

Reply:
Yeah it only works on direct OrderBy -> Where chains, if you have something in between it does not capture it. I think it would be quite messy to make it work. I also am not sure if you could say they are FNs. Logically they are, but the rule looks for invocation -> invocation. 

For example, I have this on UTs and when I was writing it I marked it as compliant.

```csharp
list.OrderBy(x => x).Select(x => x).Where(x => true); // Compliant
```

If you disagree, I can mark it as an FN, but I don't think it's something we will ever fix (it seems very hard and expensive)